Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,260 @@
+                self.num_functions, self._full_gradient_at_snapshot, 1., out=out)
+
+        return out
+
+    def _update_full_gradient_and_return(self, x, out=None):
+        """
+        Takes a "snapshot" at the point :math:`x`, saving both the point :math:` \tilde{x}=x` and its gradient :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}`. The function returns :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}` as the gradient calculation. If :code:`store_gradients==True`, the gradient of all the :math:`f_i`s is computed and stored at the "snapshot"..
+        
+        Parameters
+        ----------
+        Takes a "snapshot" at the point :math:`x`. The function returns :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}` as the gradient calculation. If :code:`store_gradients==True`, the gradient of all the :math:`f_i`s is stored, otherwise only  the sum of the gradients and the snapshot point :math:` \tilde{x}=x` are stored.
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+
+        Returns
+        --------
+        DataContainer (e.g. ImageData)
+            the value of the approximate gradient of the sum function at :code:`x` given a `function_number` in {0,...,len(functions)-1}
+        """
+
+        self._svrg_iter_number += 1 
+
+        if self.store_gradients is True:
+            if self._list_stored_gradients is None: 
+                #Save the gradient of each individual f_i and the gradient of the full sum at the point x. 
+                self._list_stored_gradients = [
+                    fi.gradient(x) for fi in self.functions]
+                self._full_gradient_at_snapshot = sum(self._list_stored_gradients, start=0*x)
+            else:
+                for i, fi in enumerate(self.functions):
+                    fi.gradient(x, out=self._list_stored_gradients[i])
+                self._full_gradient_at_snapshot.fill( sum(self._list_stored_gradients, start=0*x))
+           
+        else:
+            #Save the snapshot point and the gradient of the full sum at the point x. 
+            self._full_gradient_at_snapshot = self.full_gradient(x, out=self._full_gradient_at_snapshot) 
+            
+        if self.snapshot is None: 
+            self.snapshot = x.copy()
+
+        self.snapshot.fill(x) 
+
+        #In this iteration all functions in the sum were used to update the gradient 
+        self._update_data_passes_indices(list(range(self.num_functions))) 
+
+        #Return the gradient of the full sum at the snapshot. 
+        if out is None:
+            out = self._full_gradient_at_snapshot
+        else:
+            out.fill( self._full_gradient_at_snapshot)
+
+        return out
+
+
+class LSVRGFunction(SVRGFunction):
+    """""
+    A class representing a function for Loopless Stochastic Variance Reduced Gradient (SVRG) approximation. This is similar to SVRG, except the full gradient at a "snapshot"  is calculated at random  intervals rather than at fixed numbers of iterations. 
```

Comment:
```suggestion
    A class representing a function for Loopless Stochastic Variance Reduced Gradient (SVRG) approximation. This is similar to SVRG, except the full gradient at a "snapshot"  is calculated at random intervals rather than at fixed numbers of iterations. 
```